File: packages/capacitor-notificare-push-ui/android/src/main/java/re/notifica/push/ui/capacitor/EventBroker.kt
package re.notifica.push.ui.capacitor

import com.getcapacitor.JSObject
import org.json.JSONObject
import re.notifica.internal.NotificareLogger

internal typealias NotifyListenersFunction = (name: String, data: JSObject?) -> Unit

internal object EventBroker {
    private var notifyListenersFunction: NotifyListenersFunction? = null
    private val eventQueue = mutableListOf<Event>()

    fun setup(notifyListenersFunction: NotifyListenersFunction) {
        this.notifyListenersFunction = notifyListenersFunction

        if (eventQueue.isNotEmpty()) {
            NotificareLogger.debug("Processing event queue with ${eventQueue.size} items.")
            eventQueue.forEach { notifyListenersFunction(it.name, it.data) }
            eventQueue.clear()
        }
    }

    fun dispatchEvent(name: String, data: JSONObject?) {
        dispatchEvent(name, data?.let { JSObject.fromJSONObject(it) })
    }

    fun dispatchEvent(name: String, data: JSObject?) {
        val notifyListenersFunction = notifyListenersFunction ?: run {
            eventQueue.add(Event(name, data))
            return
        }

        notifyListenersFunction(name, data)
    }

    private data class Event(
        val name: String,
        val data: JSObject?,
    )
}

File: packages/capacitor-notificare-authentication/ios/Plugin/EventBroker.swift

import Foundation
import Capacitor
import NotificareKit

internal typealias NotifyListenersFunction = (_ eventName: String, _ data: [String: Any]?) -> Void

internal class EventBroker {
    private init() {}
    
    internal static let instance: EventBroker = .init()
    
    private var notifyListenersFunction: NotifyListenersFunction?
    private var queue:  [Event] = []
    
    func setup(_ notifyListenersFunction: @escaping NotifyListenersFunction) {
        self.notifyListenersFunction = notifyListenersFunction
        
        if !queue.isEmpty {
            NotificareLogger.debug("Processing event queue with ${eventQueue.size} items.")
            queue.forEach { dispatchEvent($0.name, data: $0.data)}
            queue.removeAll()
        }
    }
    
    internal func dispatchEvent(_ name: String, data: [String: Any]?) {
        guard let notifyListenersFunction = notifyListenersFunction else {
            queue.append(Event(name: name, data: data))
            return
        }
        
        notifyListenersFunction(name, data)
    }
    
    internal struct Event {
        let name: String
        let data: [String: Any]?
    }
}
